veps/sig-storage/incremental-backup.md

```diff
@@ -0,0 +1,362 @@
+*NOTE: In the case of a hot-plugged disk, the VM will need to be restarted for it to be included in the backup and have CBT. Once live addition of the overlay is supported, restarts will no longer be required.
+
+### Full backup
+
+Libvirt's domain commands will be used to leverage QEMU backup capabilities, with the aim of making our backup API as similar as possible to libvirt's backup and incremental backup API.
+
+To initiate a backup, a VirtualMachineBackup resource will be created. This resource will be managed by a new VMBackup controller.
+
+As outlined in Libvirt, there are two general modes for backup:
+
+```
+A `push` mode (where the hypervisor writes out the data to the destination file, which may be local or remote), and a `pull` mode (where the hypervisor creates an NBD server that a third-party client can then read as needed, and which requires the use of temporary storage, typically local, until the backup is complete).
+```
+* For the first phase of implementation, we will start with push mode. In this mode, a PVC will either be provided or be created, and will be hot-plugged to the virt-launcher. Backup will be initiated, and the data will be pushed to the attached PVC. Afterward, the PVC will be unplugged and the user can decide what to do with the it, and how to move it to a remote location if desired.
+* In the second phase, we will implement pull mode. This will involve defining a complete network API to expose the endpoint that the user can connect to in order to pull the backup. The design for this API is still **TBD** and will be updated accordingly. The endpoint information will be updated in the VMBackup CR for user reference. As specified in libvirt documentation, an additional scratch PVC, sized to accommodate all the VM disks, will be required to serve as temporary backup space to store any blocks that change during the backup job. The PVC must be large enough to hold the total size of all the VM disks, as in the worst-case scenario, the entire storage could be written to during the backup.
+
+Once the backup is initialized, the controller will pass a backup command to the virt-launcher via the virt-handler using a subresource, containing all the relevant information. Before the backup begins, an FSFreeze command will be issued to ensure file system consistency during the backup. Then, [`virDomainBackupBegin`](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainBackupBegin) will be invoked, which, as documented in libvirt, starts a point-in-time backup job for the specified disks of the running domain. This job captures the domain's disks state at the time of initiation, allowing to then call FSThaw. This minimizes guest downtime and enables the backup to be fetched while the guest continues its workload.
```

Comment:
How about we leave the pull method out of this proposal for now until details are fully fleshed out? Can be added in subsequent VEP?